tests/valgrind.supp

```diff
@@ -0,0 +1,84 @@
+{
```

Comment:
This is scary. I would like to get better explanation why do you think these are false positives.

Reply:
Indeed, these look scary.  For this reason, we spent a long long time investigating the issue, but the code seems sound.  There are two independent sets of locks, both need to be converted to mutexes to stop the helgrind warnings — it's implausible we'd make the same mistake in both cases.  All locks are taken and released in the same functions.  There's no obvious way to recurse, nor to release a lock from a different thread that took it (which is also said to cause such an error).  Thus, it _looks_ like a set of false positives.

Reply:
Ok, it makes sense now. The problem likely comes from #18 or #22 or both. This is because Helgrind/DRD doesn't track spin locks.
I think the solution to this problem is to have a compile time switch which replaces spin locks with mutexes and test that under Valgrind.

Reply:
Done

Reply:
@marcinslusarz PRs #18 or #22 are reverted.